typing/btype.ml

```diff
@@ -744,18 +745,26 @@ let last_snapshot = s_ref 0
 let log_type ty =
   if ty.id <= !last_snapshot then log_change (Ctype (ty, ty.desc))
 let link_type ty ty' =
+  (* CJC XXX look through all uses of link type and decide which ones need
```

Comment:
I think the main point is that `link_type` throws away `ty`'s description, and therefore any layout information in it.  So, for example, when we call `link_type` in the unification case where `ty` is a `Tvar`, we first make sure other thing has a layout which is a sublayout of `ty`'s.  I've added a comment to that effect.

It is challenging to add an assertion here as `link_type` is defined well before any layout checking functions.  For fun, I tried adding one just in the case where `ty` and `ty'` are both vars, since then we can see the two layouts and easily check the sublayouting.  It is violated many times while compiling stdlib - the one I looked at is a case where we're skipping the layout unification check from subst, while we're also in the delayed-layout-checks gap.  Which seems fine.